src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs

```diff
@@ -19,8 +19,10 @@ public void FinalizerRunsIfTimerAwaitableReferencesObject()
             var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
             UseTimerAwaitableAndUnref(tcs);
 
+            var timeout = tcs.Task.OrTimeout();
```

Comment:
Below, we should be verifying `Assert.True(tcs.Task.IsCompletedSuccessfully);` to verify it didn't timeout.

I wonder why we need the while loop here and not in Kestrel's HttpConnectionManagerTests. Is it something about TimerAwaitable? I think the real fix might be to set the TimerAwaitable's dueTime to a very large value so you don't risk calling GC.Collect() while the timer is executing the callback.

https://github.com/dotnet/aspnetcore/blob/100823af0d50333b6ee10147a20b35196fcbbc7d/src/Servers/Kestrel/Core/test/HttpConnectionManagerTests.cs#L25-L35

Also, why is this test tracking _count? If anything I would check that Start() never finishes meaning the _timer is never disposed.

Reply:
> Below, we should be verifying `Assert.True(tcs.Task.IsCompletedSuccessfully);` to verify it didn't timeout.

Whoops.



> I wonder why we need the while loop here and not in Kestrel's HttpConnectionManagerTests.

That test isn't doing anything async it looks like. I'm guessing that an async state machine is rooting the object so there is a race

Reply:
I answered my own question. I think the real reason is we're `calling GC.Collect() while the timer is executing the callback.` That's why I suggesting making the dueTime really large instead of adding this loop.